lib/format_parser.rb

```diff
@@ -30,24 +39,30 @@ def self.parse_http(url)
     parse(cached_io)
   end
 
-  def self.parse(io)
+  def self.parse(io, **opts, &proc)
+    config = parse_config(**opts, &proc)
     # If the cache is preconfigured do not apply an extra layer. It is going
     # to be preconfigured when using parse_http.
     io = Care::IOWrapper.new(io) unless io.is_a?(Care::IOWrapper)
+    results = []
 
     # Always instantiate parsers fresh for each input, since they might
     # contain instance variables which otherwise would have to be reset
     # between invocations, and would complicate threading situations
-    parsers = @parsers.map(&:new)
+    parsers = @parsers.select do |p|
```

Comment:
I think this block could use more lines because it does two things - it collects parsers we are willing to try and it instantiates them. That's a bit stuffed for one line. Moreover - it does selections on two criteria. Also either of these criteria can be set, while the other remains at `:all`, so it could be useful to do these selects in steps.

Reply:
I don't agree. It's doing one thing only based on two criteria. Both can be set at the same time, it's not exclusive: nobody is stopping you to do something like:

```ruby
FormatParser.parse(f, natures: [:audio], formats: [:mp3])
```

Reply:
I don't think I mentioned "exclusive". What I meant was that you can do this:

`parse(f, natures: [:audio])`

as well as

`parse(f, formats: [:png, :tif])`

as well as both. Each of these keyword arguments can be turned into a separate pass of the `.select`, which, while not as optimized, will be more readable.